### PR TITLE
Fix keyboard navigation in FlatList using XAML properties.

### LINKED ISSUE #2107: Improved keyboarding behaviors for lists
Some components don't have appropriate focus visuals and keyboarding behaviors when they receive keyboard focus. This issue is tracking using the `onKeyDown`, `onKeyUp` events to implement the right behaviors for lists and popups based on [Keyboarding interactions](https://docs.microsoft.com/en-us/windows/uwp/design/input/keyboard-interactions) guidance document for Windows. 

- [x] FlatList - Focus rect is drawn around the entire component
- [ ] FlatList - User can't move focus to a specific item using arrow keys

Related to #3006

### PR DESCRIPTION
**Pull request by rectified95**
**What**
This PR implements most of the native XAML keyboard behavior for the RN FlatList component: moving between list entries with arrows, and forward tabbing out of the list - currently pressing tab moves focus to the next list entry, while an arrow scrolls the list.
Closes #2107 

**How**
RN uses ScrollView inside FlatList, therefore we need a way of telling the ScrollViewManager whether it is rendering a FlatList or a ScrollView (the latter should not have its keyboarding behavior changed) - setting a flag in VirtualizedList.js to indicate that.

The navigation itself is achieved by setting 2 propserties on the XAML ScrollViewer:
```cpp
TabFocusNavigation(xaml::Input::KeyboardNavigationMode::Once);
XYFocusKeyboardNavigation(xaml::Input::XYFocusKeyboardNavigationMode::Enabled);
```

**Considerations**
This is done in a much simpler way than intercepting key presses, which is what #6565 tries to do.
That PR implements the FocusZone component, which could also be used to wrap a FlatList to get this behavior.
I see some issues with that: 
- I think that FlatList should act correctly OOTB, without the need to wrap it in other elements. 
- The current idea for the \<FocusZone> does not allow nested zones, ie. if there is another element in the same focus zone as the list, pressing the Down arrow, is ambiguous. 
- Lastly, the FocusZone implementation still has problems beyond the most basic scenarios due to RNW's unusually complicated UI tree.
- Keyboard behavior specific to desktop makes FlatList a more complex component compared to mobile, which I believe justifies forking the JS file behind it (it's also a minimal change, with negligible maintenance cost during upstream integrations).

**What's left**
Due to the way RNW creates the UI tree, backward tabbing with Shift+Tab still behaves like the Up arrow; this is the case both for this approach and that in #6565. Filing a bug to track fixing that, but it will require a major refactoring. I've so far debugged the XAML FocusManager and don't see a good solution yet.

**EDIT**: Found the bug in the WinUI repo: https://github.com/microsoft/microsoft-ui-xaml/issues/1363

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/6826)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/Views/ScrollViewManager.cpp
Comment (asklar) [inline vnext/Microsoft.ReactNative/Views/ScrollViewManager.cpp:226]: nit :)
```suggestion
        // We use the XAML ScrollViewer for both the RN ScrollView and FlatList - need arrow navigation in FlatList
```

File: vnext/Microsoft.ReactNative/Views/ScrollViewManager.cpp
Comment (asklar) [inline vnext/Microsoft.ReactNative/Views/ScrollViewManager.cpp:231]: `try_as` must be checked or we will null-deref

**Comment by rectified95** [comment]
Commenting to remove stale label. Will follow up on this.

**Comment by rectified95** [comment]
Had issues passing the Flow typecheck, otherwise the PR worked with the caveats listed in the description.
Colleagues at FB also used  similar workaround, so might check in with them before merging.

**Review by NickGerleman — CHANGES_REQUESTED** [review]
I think this change needs some more thought. I don't think this is the right layer in the FlatList architecture to be making it, and also worry the behavior will regress existing users.

It's forcing all users of FlatList, regardless of scenario, into new behavior. This new behavior is different from other platforms (who also have keyboard), and their expectations.

This will need to be controllable in some way. A path to start might be to expose the functionality as an opt-in prop. E.g. something like `unstable_enableXYFocusNavigation`. That could unblock folks, while we figure out the right API here, and the way to reconcile it with the more popular platforms.

**Comment by NickGerleman** [comment]
Logistically, I've been making changes to VirtualizedList over the last couple of months to unblock the commenting team. Some are in this repo, and some are coming soon into it. RNW right now is using an external package for VirtualizedList (instead of the built-in one this PR targets), but for the XAML version of RNW, I am going to move it back to how it was when this PR was first made.

**Comment by lyahdav** [comment]
> I think this change needs some more thought. I don't think this is the right layer in the FlatList architecture to be making it, and also worry the behavior will regress existing users.
> 
> It's forcing all users of FlatList, regardless of scenario, into new behavior. This new behavior is different from other platforms (who also have keyboard), and their expectations.
> 
> This will need to be controllable in some way. A path to start might be to expose the functionality as an opt-in prop. E.g. something like `unstable_enableXYFocusNavigation`. That could unblock folks, while we figure out the right API here, and the way to reconcile it with the more popular platforms.

We've found that we need XYFocusNavigation in places other than FlatList sometimes, so creating a FocusZoneView component gave us that flexibility as we can wrap any component with it. Also this makes it trivial to not have XYFocusNavigation in any FlatLists we may not want it on. Note that it's not as simple as exposing XAML's XYFocusNavigation prop because it's buggy. If you want our latest FocusZoneView code that works around all the issues we've identified, let me know.

**Comment by NickGerleman** [comment]
This sounds like a good solution. @lyahdav is the current code you have in a publicly shareable state?

**Comment by lyahdav** [comment]
> This sounds like a good solution. @lyahdav is the current code you have in a publicly shareable state?

@NickGerleman I put up a WIP PR here: https://github.com/microsoft/react-native-windows/pull/9361

A few notes:

1. I haven't tested the PR as-is. That code is lifted from a working version of FocusZoneView against our much older fork of RNW, so there may be issues.
2. In our setup, FocusZoneViewManager is in product code, so it may take some editing to get it working from within the RNW repo.
3. The PR doesn't have code to register the view manager.
4. There is XAML Islands specific code in that PR that should probably be gated with an IfIslands check.

**Comment by lyahdav** [comment]
@NickGerleman also I meant to add, it might make sense for such a component to live alongside this Mac FocusZone in FluentUI React Native: https://github.com/microsoft/fluentui-react-native/tree/master/packages/components/FocusZone/macos

**Pull request closed without merging**
